Per mod settings #8276

rob-v opened the pull request:
Added support for per mod settings in settings._mod_.yaml file. Currently supported options:

Player - Name, Color, LastServer (MP Direct IP Address:port)
Faction, Team (1 Faction for both Skirmish and Multiplayer)
Hotkeys
Server - all options from settings.yaml (MP Create Game name, port, map,...)

Commit:
Per mod settings

pchote commented on OpenRA.Game/ModData.cs:
```diff
@@ -45,6 +45,8 @@ public sealed class ModData : IDisposable
 		readonly Lazy<IReadOnlyDictionary<string, SequenceProvider>> defaultSequences;
 		public IReadOnlyDictionary<string, SequenceProvider> DefaultSequences { get { return defaultSequences.Value; } }
 
+		public IEnumerable<Traits.FactionInfo> Factions { get { return DefaultRules.Actors.Count > 0 ? DefaultRules.Actors["world"].TraitInfos<Traits.FactionInfo>() : null; } }
```
This will make it more difficult to support map customizable factions.

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -59,13 +59,13 @@ public static class Game
 		static Task discoverNat;
 
-		public static OrderManager JoinServer(string host, int port, string password, bool recordReplay = true)
+		public static OrderManager JoinServer(string host, int port, string password, bool recordReplay = true, bool isSkirmish = false)
```
I really don't like the idea of adding this, but unfortunately don't have the time to think about this and offer an alternative.

rob-v replied:
I didn't find another way how to find out if it is skirmish or MP.

pchote commented on OpenRA.Game/Settings.cs:
```diff
@@ -336,9 +336,22 @@ public class ChatSettings
 		public bool ConnectAutomatically = false;
 	}
 
+	public class SkirmishSettings
```
Why do we care about separating skirmish vs mp faction/team?

rob-v replied:
I only assume it could be useful, but it could be removed. E.g. I play MP as one faction, but I try/practice another in Skirmish.

pchote commented on OpenRA.Game/Settings.cs:
```diff
@@ -457,5 +458,51 @@ static void LoadSectionYaml(MiniYaml yaml, object section)
+		internal void LoadModSettings(string mod, string settingsFile, Arguments args)
+		{
+			modSettingsFile = settingsFile;
+			ModSections = new Dictionary<string, object>()
```
Hotkey settings are our most important mod-specific dependency.  Right now we expose hotkey configuration for all mods to all other mods, instead of letting each mod define the hotkeys that it actually uses.

rob-v replied:
Ok, I will add hotkeys

rob-v commented:
Added support for hotkeys per mod
Removed separate Faction for Skirmish/Multiplayer (I didn't like the Skirmish flag also so I agree it is now cleaner)

pchote commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -308,9 +308,9 @@ void ValidateClient(Connection newConn, string data)
 					PreferredColor = handshake.Client.Color,
 					Color = handshake.Client.Color,
-					Faction = "Random",
+					Faction = handshake.Client.Faction,
```
You'll need to validate these first to prevent malicious players from sending garbage to kill the server.

rob-v replied:
The faction is already validated after loading it from settings. Isn't it sufficient? I think also now it isn't validated each time you change Faction.

pchote commented on OpenRA.Game/Settings.cs:
```diff
@@ -336,9 +336,16 @@ public class ChatSettings
 
+	public class LobbySettings
+	{
+		public string Faction = "Random";
```
Can you please move player color here from `PlayerSettings` too?  That is closer to the faction than the player name because the mod tilesets etc restrict the color choices differently for each mod.

rob-v replied:
Color is already per mod. Player section (Name, Color, Server) is now also in mod settings, it overwrites main Player settings so it could be removed from settings. I kept it now so current user won't lose their settings. They are used first time and then saved also in mod settings to be used next time.

pchote replied:
We'll want the player name to persist across mods.

rob-v replied:
You mean you want always 1 player name for all mods?

reaperrr replied:
@pchote If we want to improve our support for parallel installations of OpenRA mods/games, I wonder if we should really make names persistent across mods. Sure, for our shipping mods it would be more convenient because players are used to it, but for 3rd-party mods not so much, and there might actually be players who want to use different names for different mods.

pchote replied:
Yes, I agree that this is an important part of the parallel installation rework, and i'm planning to put some proper thought into this and file a PR dedicated to that after #13223 and #13226 are merged.

We can then rebase this one on top, so that we don't mix fundamental infrastructure changes with new mod-level features.

reaperrr replied:
Just so it won't be forgotten, once we have all this merged, #7213 should be within grasp, which would be a significant first step for campaign support, in my opinion.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:
```diff
@@ -148,7 +148,7 @@ void ConnectionStateChanged(OrderManager om)
 			colorPreview.Color = Game.Settings.Player.Color;
 
-			foreach (var f in modRules.Actors["world"].TraitInfos<FactionInfo>())
+			foreach (var f in modData.Factions)
```
Please keep this line as it is; in the future this will change to `mapRules`.

reaperrr commented:
Needs rebase.

rob-v commented:
Rebased.

rob-v commented:
@pchote You recently rewrote mod support design, if you have an idea also how to enhance it to support these per mod settings, close this PR, might be easier.

pchote commented:
I'm planning (at some point soonish) my own take on the per-mod settings, but this is still useful to be rebased on top of that for the new settings.

rob-v commented:
Per mod settings (planned new PR) and new settings (faction,...) are different things. This PR combines both in 1 commit. First/major part of this PR is obsolete, so better to close it. The new PR can add also a setting or it can be another PR to extend it.
